File: problems/009-sparse-file-quota/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	var f, n int
	var cap uint64
	if _, e := fmt.Fscan(in, &f, &n, &cap); e != nil {
		return
	}
	size := make([]uint64, f+1)
	cur := make([]uint64, f+1)
	var used, peak uint64
	for ; n > 0; n-- {
		var op string
		var x int
		var v uint64
		fmt.Fscan(in, &op, &x, &v)
		err := false
		if op == "SEEK" {
			cur[x] = v
		} else {
			ns := v
			if op == "WRITE" {
				ns = size[x]
				if v > 0 && cur[x]+v > ns {
					ns = cur[x] + v
				}
			}
			if ns > size[x] && ns-size[x] > cap-used {
				err = true
			} else {
				if ns >= size[x] {
					used += ns - size[x]
				} else {
					used -= size[x] - ns
				}
				size[x] = ns
				if op == "WRITE" && v > 0 {
					cur[x] += v
				}
			}
		}
		if used > peak {
			peak = used
		}
		if err {
			fmt.Fprint(out, "ERR QUOTA")
		} else {
			fmt.Fprint(out, "OK")
		}
		fmt.Fprintln(out, "", size[x], cur[x], used)
	}
	fmt.Fprintln(out, "SUMMARY", used, peak)
}

File: problems/009-sparse-file-quota/solutions/cpp/main.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>
using namespace std;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int f, n;
  uint64_t cap;
  if (!(cin >> f >> n >> cap))
    return 0;
  vector<uint64_t> sz(f + 1), cur(f + 1);
  uint64_t used = 0, peak = 0;
  while (n--) {
    string op;
    int x;
    uint64_t v;
    cin >> op >> x >> v;
    bool err = false;
    if (op == "SEEK")
      cur[x] = v;
    else {
      uint64_t ns = op == "WRITE" ? (v ? max(sz[x], cur[x] + v) : sz[x]) : v;
      if (ns > sz[x] && ns - sz[x] > cap - used)
        err = true;
      else {
        if (ns >= sz[x])
          used += ns - sz[x];
        else
          used -= sz[x] - ns;
        sz[x] = ns;
        if (op == "WRITE" && v)
          cur[x] += v;
      }
    }
    peak = max(peak, used);
    cout << (err ? "ERR QUOTA" : "OK") << ' ' << sz[x] << ' ' << cur[x] << ' '
         << used << '\n';
  }
  cout << "SUMMARY " << used << ' ' << peak << '\n';
}
